pkg/workspace/repository.go

```diff
@@ -0,0 +1,33 @@
+package workspace
+
+import (
+	"github.com/odpf/siren/domain"
+	"github.com/pkg/errors"
+)
+
+type Repository struct {
+	Slacker domain.SlackService
+}
+
+func NewRepository(service domain.SlackService) SlackRepository {
+	return &Repository{
+		Slacker: service,
+	}
+}
+
+func (r Repository) GetWorkspaceChannel(token string) ([]Channel, error) {
+	r.Slacker.UpdateClient(token)
```

Comment:
- This would update the existing instance of `r` and cause issues with multiple requests for multiple entities using this simultaneously. Basically, I see that a singleton instance of `SlackNotifierClient` is being updated while it might be in use by other calls.
- `slack.New` is creating a new object of the client. This would create that object on each call. Thus it would increase memory usage tremendously on every call.

Reply:
We can't create a singleton class for slack for sure because we need to change the slack client based on request types.
One thing which we can implement to stop the creation of new objects on each call is a map
```
type ClientService struct {
	clients     map[string]SlackCaller
	SlackClient SlackCaller
}

func (c *ClientService) UpdateClient(token string) {
	client, ok := c.clients[token]

	if !ok {
		client := slack.New(token)
		c.clients[token] = client
	}
	c.SlackClient = client
}
```
And to stop interference between multiple requests for multiple entities we can use mutex during the caller
```
mu      sync.Mutex
```

but I will discuss with @whoAbhishekSah to come up with a better approach though.

Reply:
We can try this out and in order to test it better can we write integration test with multiple requests coming in for different tokens?

Reply:
@AkarshSatija  we can simply use Value Receiver in SlackCaller methods. e.g. changing this 
`func (c *ClientService) UpdateClient(token string)` to `func (c ClientService) UpdateClient(token string) ` would ensure the ClientService is being copied(by value) on every call, resulting in safe concurrent assignment.

Reply:
We can add one test for ensuring safe concurrent access to update the slack client on each request.